fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/lock/InMemoryResourceLockManager.java

```diff
@@ -29,67 +35,86 @@ public class InMemoryResourceLockManager implements ResourceLockManager {
 
     private static final Logger LOG = LoggerFactory.getLogger(InMemoryResourceLockManager.class);
 
-    private final Map<String, Set<String>> transactionLocks;
-    private final Set<String> lockedResources;
-    private final Map<String, Object> internalResourceLocks;
+    private final Map<String, Set<ResourceLock>> transactionLocks;
+    private final Map<String, Set<ResourceLock>> internalResourceLocks;
+    private final Map<String, Set<ResourceLock>> resourceLocks;
 
     public InMemoryResourceLockManager() {
         transactionLocks = new ConcurrentHashMap<>();
-        lockedResources = Sets.newConcurrentHashSet();
+        resourceLocks = new ConcurrentHashMap<>();
         internalResourceLocks = Caffeine.newBuilder()
                 .expireAfterAccess(10, TimeUnit.MINUTES)
-                .<String, Object>build()
+                .<String, Set<ResourceLock>>build()
                 .asMap();
     }
 
     @Override
-    public void acquire(final String txId, final FedoraId resourceId) {
+    public void acquire(final String txId, final FedoraId resourceId, final ResourceLockType lockType) {
         final var resourceIdStr = resourceId.getResourceId();
 
-        if (transactionHoldsLock(txId, resourceIdStr)) {
+        if (transactionHoldsExclusiveLock(txId, resourceIdStr)) {
             return;
         }
 
-        synchronized (acquireInternalLock(resourceIdStr)) {
-            if (transactionHoldsLock(txId, resourceIdStr)) {
+        synchronized (acquireInternalLock(txId, resourceIdStr, lockType)) {
+            if (transactionHoldsExclusiveLock(txId, resourceIdStr)) {
                 return;
             }
 
-            if (lockedResources.contains(resourceIdStr)) {
-                throw new ConcurrentUpdateException(
-                        String.format("Cannot update %s because it is being updated by another transaction.",
-                                resourceIdStr));
+            if (resourceLocks.containsKey(resourceIdStr)) {
+                final var locks = resourceLocks.get(resourceIdStr);
+                if (locks.size() > 0 && (
+                        lockType.equals(EXCLUSIVE) && locks.parallelStream().anyMatch(l ->
```

Comment:
I'm a little dubious about the use of `parallelStream()`. That will farm the work out to a shared fork-join pool, which I don't think will necessarily speed things up, especially if the list is relatively small.

Shouldn't there be parens around `lockType.equals(EXCLUSIVE) && locks.parallelStream().anyMatch(l -> !l.getTransactionId().equals(txId))`? That said, do you even need to do the search because you already know the transaction doesn't have a lock on the resource?

Reply:
You are correct on both counts, it is not wrapped correctly and it is redundant.

Reply:
Sorry just working this through. So my thought process here is 
* if the current TX has an exclusive lock then you don't get here.
* if you want an exclusive lock and there is already a non-exclusive lock held by a different transaction, then you need to fail. 
* If the current TX has a non-exclusive lock, then we can grab an exclusive lock on the same resource.

I can make it (hopefully) easier to read like this?
```
if (locks.size() > 0 &&
    (
        // Can't get an exclusive lock on a resource that already has any lock not owned by the 
        // current transaction. If the current transaction held the exclusive lock we would 
        // have exited above.
        (lockType.equals(EXCLUSIVE) && 
                     locks.parallelStream().anyMatch(l ->!l.getTransactionId().equals(txId))
        ) ||
        // Can get any lock if there is an exclusive lock, again the current transaction
        // would not own it or we would have exited above.
        locks.parallelStream().anyMatch(l -> l.hasLockType(EXCLUSIVE))
     )
) {
```

Reply:
Ah, yes, you're right. The check on line 60 was looking specifically for an exclusive, so your logic here makes sense.